docs(release-note): add databend changelog 1.2 (#11925)
datafuselabs/databend: website/blog/2023-06-29-databend-v1.2-release.md

In addition, we have designed and implemented enterprise features such as **Computed Columns**, **VACUUM TABLE**, and **Serverless Background Service**. If you are interested in these features, you can contact [Databend Support](https://www.databend.com/contact-us) for upgrade information or visit [Databend Cloud](http://app.databend.cn/) for an instant experience.

## New Features and Enhancements

Discover Databend's new features and find your perfect fit with a quick overview.

### New Data Type: `BITMAP`

Databend has added support for the `BITMAP` datatype and implemented a series of related functions.

`BITMAP` is a type of compressed data structure that can be used to efficiently store and manipulate sets of boolean values. It has wide applications in data analysis and querying, providing fast set operations and aggregation capabilities. Common use cases include Distinct Count, Filtering & Selection, and Compressed Storage.

> Our implementation of the BITMAP data type utilizes `RoaringTreemap`. Using this data structure brought us improved performance and decreased memory usage in comparison to alternative bitmap implementations.

Reviewer: ```suggestion
Our implementation of the BITMAP data type utilizes `RoaringTreemap`, which brings us improved performance and decreased memory usage in comparison to other bitmap implementations.
```

```sql
SELECT user_id, bitmap_count(page_visits) AS total_visits
FROM user_visits

+--------+------------+
|user_id |total_visits|
+--------+------------+
|       1|           4|
|       2|           3|
|       3|           4|
+--------+------------+
```

If you are interested in learning more, please check out the resources listed below:

- [Docs |  Data Types - BITMAP](https://databend.rs/doc/sql-reference/data-types/data-type-bitmap)
- [Docs | SQL Functions - Bitmap Functions](https://databend.rs/doc/sql-functions/bitmap-functions/)
- [Paper | Consistently faster and smaller compressed bitmaps with Roaring](https://arxiv.org/pdf/1603.06549.pdf)

### Direct Query of CSV/TSV/NDJSON Files Using Column Position

> To query files such as CSV/TSV/NDJSON that do not have a schema, they had to be loaded into a table before querying. However, sometimes users do not know the specific details of the file in advance (such as how many columns a CSV file has), or they just want to make temporary queries.

Reviewer: ```suggestion
To query files such as CSV/TSV/NDJSON that do not have a schema, they have to be loaded into a table before querying. However, sometimes you do not know the specific details of the file in advance (such as how many columns a CSV file has), or they just want to make temporary queries.
```

> To address this, Databend introduces **Column Position** and uses the `$N` syntax to represent the Nth column. All columns in CSV/TSV files are treated as `String` type. If the number of columns in a row is less than the column number used, it will be padded with an empty string. NDJSON files have only one column `$1`, which is of type `Variant` .

Reviewer: ```suggestion
To make it easier, Databend introduces **Column Position** and uses the `$N` syntax to represent the Nth column. All columns in CSV/TSV files are treated as `String` type. If the number of columns in a row is less than the column number used, it will be padded with an empty string. NDJSON files have only one column `$1`, which is of type `Variant` .
```

> When combined with the `COPY` statement, this capability can be used to load only the required columns and perform data conversion using functions during loading.

Reviewer: ```suggestion
This capability allows you to selectively load specific columns using the COPY INTO command. Additionally, you can perform data conversion using functions during the loading process.
```

```sql
SELECT $1 FROM @my_stage (FILE_FORMAT=>'ndjson')

COPY INTO my_table FROM (SELECT TRIM($2) SELECT @my_stage t) FILE_FORMAT = (type = CSV)
```

If you are interested in learning more, please check out the resources listed below:

- [Docs | SELECT - Column Position](https://databend.rs/doc/sql-commands/query-syntax/query-select#column-position)

> ### Designed and Implemented a New Hash Table to Improve Hash Join Performance

Reviewer: ```suggestion
### New Hash Table: Improved Hash Join Performance
```

> In the past, Databend's hash table was specifically designed to meet the needs of aggregation operators. To further improve Hash Join performance, we have designed and implemented a new hash table optimized for Hash Join. Through parallelized design, Databend can fully utilize computing resources while also becoming more precise in memory control, avoiding unnecessary memory overhead and significantly improving Hash Join performance.

Reviewer: ```suggestion
Initially, Databend's Hash table was specifically designed for aggregation operators. To further improve the Hash Join performance, we have designed and implemented a new Hash table optimized for Hash Join. Through parallelized design, Databend can now fully utilize computing resources while also becoming more precise in memory control, avoiding unnecessary memory overhead and significantly improving Hash Join performance.
```

Business intelligence analyst Mimoune Djouallah commented that Databend performs well and runs TPCH-SF10 in just 25 seconds under 8 cores and 32GB of memory. He also wrote a blog post titled *[Databend and the rise of Data warehouse as a code](https://datamonkeysite.com/2023/05/22/databend-and-the-rise-of-data-warehouse-as-a-code/)* .

![](../static/img/blog/databend-1-2-release/changelog-1-2-01.png)

If you are interested in learning more, please check out the resources listed below:

- [Data Infra #12 | Databend's New Hash Table for Hash Join](https://www.bilibili.com/video/BV1zu4y1Z7VA/)
- [PR #11140 | feat(query): new hash table and parallel finalize for hash join](https://github.com/datafuselabs/databend/pull/11140)

### AI Functions

Databend introduced powerful AI capabilities in version v1.2.0, achieving seamless integration of Data and AI. We can use SQL to achieve:

1. Generating SQL with Natural Language
2. Embedding Vectorization
3. Similarity Calculation
4. Text Generation

> #### Generating SQL with natural language.

Reviewer: ```suggestion
#### Generating SQL from Natural Language
```

## Growing Ecosystem

Databend's ecosystem has been further improved. It's time to introduce Databend into your data insight workflow!

### Bind `databend` into Python

Databend now offers a Python binding that allows users to execute SQL queries against Databend using Python even without deploying a Databend instance.

To use this functionality, simply import `SessionContext` from `databend` module and create an instance of it:

```python
from databend import SessionContext

ctx = SessionContext()
df = ctx.sql("select number, number + 1, number::String as number_p_1 from numbers(8)")
```

The resulting DataFrame can be converted to PyArrow or Pandas format using the `to_py_arrow()` or `to_pandas()` methods respectively:

```python
df.to_pandas() # Or, df.to_py_arrow()
```

Feel free to integrate it with your data science workflow.

- [databend · PyPI](https://pypi.org/project/databend/)

### BendSQL - Databend Native Command Line Tool

> BendSQL is a command line tool that has been designed specifically for Databend. It has been rewritten using Rust, which will not only provide support for **REST API** but also **Flight SQL**.

Reviewer: ```suggestion
BendSQL is a command-line tool specifically designed for Databend. It has been rewritten using Rust, which not only adds support for the REST API but also introduces Flight SQL functionality.
```

- [Docs | Loading Data with Tools - Flink CDC](https://databend.rs/doc/load-data/load-db/flink-cdc)

#### Tableau

Tableau is a popular data visualization and business intelligence tool. It provides an intuitive, interactive way to explore, analyze, and present data, helping users better understand the meaning and insights of their data.

Refer to [Other Databases (JDBC)](https://help.tableau.com/current/pro/desktop/en-us/examples_otherdatabases_jdbc.htm) and add `databend-jdbc` to the Tableau driver path in order to analyze data from Databend using Tableau.

![](../static/img/blog/databend-1-2-release/changelog-1-2-07.png)

If you are interested in learning more, please check out the resources listed below:

- [Blog | Databend x Tableau](https://databend.rs/blog/2023-06-01-tableau)

## Download Databend v1.2.0

If you are interested in Databend v1.2.0, go to <https://github.com/datafuselabs/databend/releases/tag/v1.2.0-nightly> to view all changelogs or download the release.

If you are using an old version of Databend, we recommend you upgrade to the latest version. For the upgrade process, please refer to:  <https://databend.rs/doc/operations/upgrade> .

## Feedback

> If you need any help when working with the new release, submit an issue on GitHub Issue.

Reviewer: ```suggestion
If you need help when working with the new release, submit an issue on GitHub Issues.
```
